text/0374-jsii-ts-version.md

```diff
@@ -0,0 +1,324 @@
+
+### How will we maintain backwards compatibility?
+
+The change only affects the [TypeScript] compiler version used internally by
+`jsii` and the versioning scheme for the `jsii` package itself.
+
+The compiler will continue to emit `.jsii` assemblies that conform to the schema
+defined in the `@jsii/spec` package, which will hence continue to be compatible
+with all other tools part of the jsii toolchain (including `jsii-pacmak`, ...).
+
+In order to maximize compatibility between [TypeScript] compiler versions, and
+since [TypeScript] occasionally introduces backwards-incompatible syntax changes
+(additions, modifications) to the declarations files (`.d.ts`), the `jsii`
+compiler will proactively produce down-leveled declarations files targeting
+[TypeScript] compiler versions used by previous (not yet *end-of-life*) releases
+of `jsii`. This can be achieved using the [`downlevel-dts`][downlevel-dts]
+utility.
+
+### Why are we doing this?
+
+Before this change, `jsii` users had been stuck with [TypeScript] `3.9` for a
+very long time, due to the introduction of several breaking changes in the
+language specification: upgrading the [TypeScript] compiler that `jsii` builds
+on would cause existing code to break.
+
+This could not be addressed by issuing a new major release of `jsii`, as this
+would cause the ecosystem to be fragmented between packages that have migrated
+to the new major release, and those that haven't. Additionally, a package could
+only migrate if all of their dependencies are already using the new `jsii` major
+release, and they'd likely need to also release a new major version as a
+consequence.
+
+In order to allow each package author to decide for themselves without hinging
+on their dependencie's choices, or influencing their dependents, we are making
+it possible for every developer to decide when they want to migrate to a new
+version of the [TypeScript] language, by following [TypeScript]'s versioning for
+the compiler, and continuing to perform coordinated releases of every other
+package in the toolchain.
+
+### Why should we _not_ do this?
+
+[Semantic versioning][semver] is the dominant versioning scheme in the
+JavaScript world, and diverging from it might break assumptions customers make
+when consuming packages from [npm](npmjs.com).
+
+Following [TypeScript] versions means releasing new `major.minor` releases (that
+may include breaking changes) more often than is currently done (the
+[TypeScript] maintainers typically declare a new `major.minor` line four
+times per year: in February, May, August and November respectively). This is a
+lot more release lines than other AWS products offer, and it would not be
+reasonable to uphold the [AWS SDKs and Tools maintenance policy][aws-policy],
+as it requires offering full support (features and bug fixes) for "Generally
+Available" releases for a minimum of 24 months. Diverging from this
+[policy][aws-policy] is a significant change that customers may not expect.
```

Comment:
Ultimately the digression from the policy is more about refusing to backport new features to jsii 3.9 than it is about changing the version number. Might be worth being more explicit about that.

Reply:
You seem to be under-estimating the maintenance burden here...

There is one new release of TypeScript per quarter (approximately), and the AWS SDKs and Tools maintenance policy mandates that everything that was deemed "GA" has full support for at least 24 months.

This does not mean "just" back-porting to jsii 3.9... Assuming we had followed this versioning strategy 2 years ago now, we'd have to back-port features & bug fixes to:
- 3.9 (at least until mid-November 2022)
- 4.0
- 4.1
- 4.2
- 4.3
- 4.4
- 4.5
- 4.6 (would be current, and 4.7 will arrive "imminently")

We'd have to build, test and release those 8 different versions.

One could argue that back-porting could be largely automated, but this fails to account for how the compiler API has significant changes in between these releases... That sounds like A LOT of thinks to maintain at the same time, and I'd rather not have to manage that many.